crates/fluvio/src/producer/mod.rs

```diff
@@ -29,355 +35,301 @@ pub struct TopicProducer {
+///     ..Default::default()
+/// };
+///
+/// // Use config on this call:
+/// fluvio.topic_producer_with_config("topic", config).await?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct ProducerConfig {
+    pub batch_duration: Duration,
+    pub batch_size: usize,
+}
+
+impl Default for ProducerConfig {
+    fn default() -> Self {
+        Self {
+            batch_duration: Duration::from_millis(DEFAULT_BATCH_DURATION_MS),
+            batch_size: DEFAULT_BATCH_SIZE_BYTES,
+        }
+    }
+}
+
+/// A unique identifier assigned to each Record sent to the Producer.
+///
+/// `RecordUid` is used to associate each input record with the eventual
+/// success or failure status of the request the record belongs to.
+/// Each call to `producer.send` returns a unique `RecordUid`, and each
+/// request status includes the `RecordUid`s of all included records.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct RecordUid(u64);
+
+/// A monotonically increasing counter for assigning Uids to records
+pub(crate) struct RecordUidGenerator {
+    next_uid: std::sync::atomic::AtomicU64,
+}
+
+impl RecordUidGenerator {
+    pub fn new() -> Self {
+        Self {
+            next_uid: std::sync::atomic::AtomicU64::new(0),
+        }
+    }
+
+    /// Claim the next UID in the monotonically increasing sequence
+    pub fn claim(&self) -> RecordUid {
+        let uid = self
+            .next_uid
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+        RecordUid(uid)
+    }
+}
+
+/// A handle for `TopicProducer`s to communicate with the shared `ProducerDispatcher`.
+///
+/// When all `TopicProducer`s are dropped, the `ProducerInner` will drop and
+/// notify the `ProducerDispatcher` to shutdown.
+pub(crate) struct ProducerInner {
+    /// A channel for sending messages to the shared Producer Dispatcher.
+    pub(crate) dispatcher: Sender<DispatcherMsg>,
+    /// An event that can notify the Dispatcher to shutdown.
+    pub(crate) dispatcher_shutdown: StickyEvent,
+    /// Generate monotonically-increasing `RecordUid`s.
+    pub(crate) uid_generator: RecordUidGenerator,
+    /// Used by the dispatcher to broadcast batch status notifications.
+    // We hold this because we cannot clone a broadcast::Receiver,
+    // we have to "subscribe" to a sender to get it
+    pub(crate) _status_sender: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Used by the producer to return batch statuses to the caller.
+    pub(crate) status_receiver: Mutex<tokio::sync::broadcast::Receiver<BatchStatus>>,
+}
+
+impl Drop for ProducerInner {
+    fn drop(&mut self) {
+        debug!("Notify dispatcher shutdown");
+        self.dispatcher_shutdown.notify();
+    }
 }
 
 impl TopicProducer {
-    pub(crate) fn new(topic: String, pool: Arc<SpuPool>) -> Self {
+    pub(crate) fn new(topic: String, pool: Arc<SpuPool>, config: ProducerConfig) -> Self {
         let partitioner = Box::new(SiphashRoundRobinPartitioner::new());
```

Comment:
What do you have in mind when you say "buffering strategy"? Is it a subset of the config options? Or is it more like an algorithm that could be represented by a trait?

Reply:
In future, we probably need support custom buffer/partitioning, so instead of harding coding this, it could be model as Trait.  Or at least minimum, tie configuration with Buffering so it's easier to handle that as separate configurable entity

Reply:
Partitioning does already have a trait that allows for it to have alternate implementations, it is just not public yet. The only thing hardcoded here is which Partitioner implementation is used (the SiphashRoundRobinPartitioner). I would say this is not part of the scope for this PR, since this is how partitioning worked before.

As far as representing the "batching strategy" as a trait, I think that could potentially be much more effort, and I think it would be best to defer that until we actually need multiple batching strategies and we know what our needs are, rather than trying to include it here where we would be guessing our future needs.